iocore/net/ProxyProtocol.cc

```diff
@@ -0,0 +1,149 @@
+/** @file
+ *
+ *  PROXY protocol definitions and parsers.
+ *
+ *  @section license License
+ *
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "ts/ink_assert.h"
+#include "ProxyProtocol.h"
+#include "I_NetVConnection.h"
+
+bool
+ssl_has_proxy_v1(NetVConnection *sslvc, char *buffer, int64_t *r)
+{
+  int64_t nl = 0;
+  char *nlp  = nullptr;
+  if (0 == memcmp(PROXY_V1_CONNECTION_PREFACE, buffer, PROXY_V1_CONNECTION_PREFACE_LEN)) {
+    nlp = (char *)memchr(buffer, '\n', PROXY_V1_CONNECTION_HEADER_LEN_MAX);
+    if (nlp) {
+      nl = (int64_t)(nlp - buffer);
+      Debug("ssl", "Consuming %" PRId64 " characters of the PROXY header from %p to %p", nl + 1, buffer, nlp);
+      char local_buf[PROXY_V1_CONNECTION_HEADER_LEN_MAX + 1];
+      memcpy(local_buf, buffer, nl + 1);
+      memmove(buffer, buffer + nl + 1, nl + 1);
+      *r -= nl + 1;
+      if (*r <= 0) {
+        *r = -EAGAIN;
+      }
+      return (proxy_protov1_parse(sslvc, local_buf));
+    }
+  }
+  return false;
+}
+
+bool
+http_has_proxy_v1(IOBufferReader *reader, NetVConnection *netvc)
+{
+  char buf[PROXY_V1_CONNECTION_HEADER_LEN_MAX + 1];
+  char *end;
+  ptrdiff_t nbytes;
+
+  end    = reader->memcpy(buf, sizeof(buf), 0 /* offset */);
+  nbytes = end - buf;
+
+  // Client must send at least 15 bytes to get a reasonable match.
+  if (nbytes < (long)PROXY_V1_CONNECTION_HEADER_LEN_MIN) {
+    return false;
+  }
+
+  if (0 != memcmp(PROXY_V1_CONNECTION_PREFACE, buf, PROXY_V1_CONNECTION_PREFACE_LEN)) {
+    return false;
+  }
+
+  int64_t nl;
+  nl = reader->memchr('\n', strlen(buf), 0);
+
+  char local_buf[PROXY_V1_CONNECTION_HEADER_LEN_MAX + 1];
+  reader->read(local_buf, nl + 1);
+
+  // Now that we know we have a valid PROXY V1 preface, let's parse the
+  // remainder of the header
+
+  return (proxy_protov1_parse(netvc, local_buf));
+}
+
+bool
+proxy_protov1_parse(NetVConnection *netvc, char *buf)
+{
+  std::string src_addr_port;
```

Comment:
Not convinced this is a good use of std::string. Is this in the critical path? I.e. if this happens on every request, makes this a sufficiently large char[] on the stack, and use that. You can use snprintf() into that safely (and I assume the size of these port strings are of known, limited size.

I do feel strongly about this on the critical path, where we avoid using std::string intentionally. If it's not in the critical path, e.g. in some configuration code that only happens infrequent, it's fine.

Reply:
Heh. This is a great place for `TextView`. Let me rip out some code for it.

Reply:
I got a bit carried away, but [here](https://github.com/SolidWallOfCode/trafficserver/blob/pr-3123/iocore/net/ProxyProtocol.cc) a suggested version of `ProxyProtocol.cc` which addresses @zwoop 's concern.